File: apps/watch/Sources/WatchCore/AliveLoop.swift
import Foundation

public enum AliveLoopError: Error, Equatable {
    case invalidSnoozeHours
}

public final class AliveLoop {
    private let config: LoopConfig
    private let decisionEngine: DecisionEngine

    private var nextAlertAt: Date?
    private var snoozeUntil: Date?

    public init(
        config: LoopConfig = LoopConfig(),
        decisionEngine: DecisionEngine = DecisionEngine()
    ) {
        self.config = config
        self.decisionEngine = decisionEngine
    }

    public func bootstrap(startAt: Date) {
        nextAlertAt = startAt.addingTimeInterval(config.intervalSeconds)
    }

    public func currentNextAlertAt() -> Date? {
        nextAlertAt
    }

    public func currentSnoozeUntil() -> Date? {
        snoozeUntil
    }

    public func applySnooze(now: Date, hours: Int) throws {
        guard [1, 2, 3].contains(hours) else {
            throw AliveLoopError.invalidSnoozeHours
        }
        snoozeUntil = now.addingTimeInterval(config.intervalSeconds * Double(hours))
    }

    public func tick(now: Date, isSleeping: Bool, isCharging: Bool) -> CheckEvent? {
        guard let scheduledAt = nextAlertAt, now >= scheduledAt else {
            return nil
        }

        let state = WatchState(
            now: now,
            isSleeping: isSleeping,
            isCharging: isCharging,
            snoozeUntil: snoozeUntil
        )

        let decision = decisionEngine.decide(state)
        let event: CheckEvent

        switch decision {
        case .sendAlert:
            event = CheckEvent(type: .alertSent, occurredAt: now)
        case .skip(let reason):
            event = CheckEvent(
                type: mapSkipReasonToEventType(reason),
                occurredAt: now,
                metadata: ["reason": reason.rawValue]
            )
        }

        nextAlertAt = now.addingTimeInterval(config.intervalSeconds)
        if let snoozeUntil, now >= snoozeUntil {
            self.snoozeUntil = nil
        }

        return event
    }

    public func onAlertDismissed(now: Date) -> CheckEvent {
        CheckEvent(type: .alive, occurredAt: now)
    }

    private func mapSkipReasonToEventType(_ reason: SkipReason) -> EventType {
        switch reason {
        case .snoozed: return .skipSnoozed
        case .sleeping: return .skipSleeping
        case .charging: return .skipCharging
        }
    }
}

File: apps/watch/src/main/kotlin/io/deathproject/watch/core/AliveLoop.kt
package io.deathproject.watch.core

import java.time.Instant

class AliveLoop(
    private val config: LoopConfig = LoopConfig(),
    private val decisionEngine: DecisionEngine = DecisionEngine(),
) {
    private var nextAlertAt: Instant? = null
    private var snoozeUntil: Instant? = null

    fun bootstrap(startAt: Instant) {
        nextAlertAt = startAt.plus(config.interval)
    }

    fun getNextAlertAt(): Instant? = nextAlertAt

    fun getSnoozeUntil(): Instant? = snoozeUntil

    fun applySnooze(now: Instant, hours: Int) {
        require(hours in setOf(1, 2, 3)) { "hours must be one of 1,2,3" }
        snoozeUntil = now.plus(config.interval.multipliedBy(hours.toLong()))
    }

    fun tick(now: Instant, isSleeping: Boolean, isCharging: Boolean): CheckEvent? {
        val scheduledAt = nextAlertAt ?: return null
        if (now.isBefore(scheduledAt)) {
            return null
        }

        val state = WatchState(
            now = now,
            isSleeping = isSleeping,
            isCharging = isCharging,
            snoozeUntil = snoozeUntil,
        )

        val event = when (val decision = decisionEngine.decide(state)) {
            Decision.SendAlert -> CheckEvent(type = EventType.ALERT_SENT, occurredAt = now)
            is Decision.Skip -> CheckEvent(
                type = skipReasonToEventType(decision.reason),
                occurredAt = now,
                metadata = mapOf("reason" to decision.reason.name.lowercase()),
            )
        }

        nextAlertAt = now.plus(config.interval)
        if (snoozeUntil != null && !now.isBefore(snoozeUntil)) {
            snoozeUntil = null
        }

        return event
    }

    fun onAlertDismissed(now: Instant): CheckEvent {
        return CheckEvent(type = EventType.ALIVE, occurredAt = now)
    }

    private fun skipReasonToEventType(reason: SkipReason): EventType {
        return when (reason) {
            SkipReason.SNOOZED -> EventType.SKIP_SNOOZED
            SkipReason.SLEEPING -> EventType.SKIP_SLEEPING
            SkipReason.CHARGING -> EventType.SKIP_CHARGING
        }
    }
}
